tests/ComApartmentVariableTests.cpp

```diff
@@ -0,0 +1,244 @@
+    REQUIRE(v == 43);
+
+    wil::reset_for_current_com_apartment(fn);
+    wil::reset_for_current_com_apartment(fn);
+    REQUIRE(wil::apartment_variable_count() == 0);
+
+    // No variable so nop
+    wil::reset_for_current_com_apartment(fn2);
+
+    // variable not present, so this has no effect
+    wil::reset_for_current_com_apartment(fn2, 42);
+    v = wil::get_for_current_com_apartment(fn2);
+    REQUIRE(v == 43);
+
+    REQUIRE(wil::apartment_variable_count() == 1);
+
+    wil::reset_for_current_com_apartment(fn2);
+    REQUIRE(wil::apartment_variable_count() == 0);
+
+    v = wil::get_for_current_com_apartment(fn2);
+    wil::reset_for_current_com_apartment(fn2, 44);
+    REQUIRE(wil::apartment_variable_count() == 1);
+
+    v = wil::get_for_current_com_apartment(fn2);
+    REQUIRE(wil::apartment_variable_count() == 1);
+    REQUIRE(v == 44);
+}
+
+struct mock_platform
+{
+    static unsigned long long GetApartmentId()
+    {
+        APTTYPE aptType;
+        APTTYPEQUALIFIER aptQualifer;
+        FAIL_FAST_IF_FAILED(CoGetApartmentType(&aptType, &aptQualifer)); // ensure COM is inited
+
+        // incorrect, but close enough for testing
+        return GetCurrentThreadId();
+    }
+
+    inline static std::unordered_map<unsigned long long, std::vector<wil::com_ptr<IApartmentShutdown>>> m_observers;
+
+    static void RegisterForApartmentShutdown(IApartmentShutdown* observer)
+    {
+        const auto id = GetApartmentId();
+        auto apt_observers = m_observers.find(id);
+        if (apt_observers == m_observers.end())
+        {
+            m_observers.insert({ id, { observer} });
+        }
+        else
+        {
+            apt_observers->second.push_back(observer);
+        }
+    }
+
+    // This is needed to simulate the platform for unit testing.
+    static auto CoInitializeExForTesting(DWORD coinitFlags = 0 /*COINIT_MULTITHREADED*/)
+    {
+        return wil::scope_exit([aptId = GetCurrentThreadId(), init = wil::CoInitializeEx(coinitFlags)]()
+        {
+            const auto id = GetApartmentId();
+            auto apt_observers = m_observers.find(id);
+            if (apt_observers != m_observers.end())
+            {
+                const auto& observers = apt_observers->second;
+                for (auto& comptr : observers)
+                {
+                    comptr->OnUninitialize(id);
+                }
+            }
+        });
+    }
+};
+
+
+wil::apartment_variable<int, mock_platform> g_v1; // 00007FF7091E9040
+wil::apartment_variable<int, mock_platform> av1, av2;
+
+TEST_CASE("ComApartmentVariable::VerifyApartmentVariable", "[com][apartment_variable]")
+{
+    auto coUninit = mock_platform::CoInitializeExForTesting(COINIT_MULTITHREADED);
+
+    std::ignore = g_v1.get_or_create(fn);
+
+    wil::apartment_variable<int, mock_platform> v1; // 00000020E63CF6F4, should be a global
+
+    REQUIRE(v1.get_if() == nullptr);
+    REQUIRE(v1.get_or_create(fn) == 42);
+    v1.set(43);

```

Comment:
Need to test setting by lvalue.

```cpp
int value = 44;
v1.set(value);
```

would have caught incorrect universal reference.